File: Problems/OneEditDistance.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    bool isOneEditDistance(string s, string t)
    {
        int ns = s.size();
        int nt = t.size();
        // ensure nt is longer string
        if (ns > nt)
            return isOneEditDistance(t, s);
        if (ns + 1 < nt || s == t)
            return 0;
        int i;
        for (i = 0; i < ns; i++)
        {
            char cs = s.at(i);
            char ct = t.at(i);
            if (cs != ct && ns == nt)
            {
                s.erase(i, 1);
                t.erase(i, 1);
                return s == t;
            }
            else if (cs != ct && ns != nt)
            {
                string tTemp = t;
                tTemp.erase(i, 1);
                if (s == tTemp)
                    return 1;
            }
        }
        t.erase(i, 1);
        return s == t;
    }

    bool isOneEditDistance2(string s, string t)
    {
        int ns = s.size();
        int nt = t.size();
        // ensure nt is longer
        if (ns > nt)
            return isOneEditDistance(t, s);
        for (int i = 0; i < ns; i++)
        {
            if (s[i] != t[i])
            {
                if (ns == nt)
                    return s.substr(i + 1) == t.substr(i + 1);
                return s.substr(i) == t.substr(i + 1);
            }
        }
        return ns + 1 == nt;
    }
};

File: Problems/OneEditDistance.py
# Given two strings s and t, return true if they are both one edit distance apart, otherwise return false.

# A string s is said to be one distance apart from a string t if you can:

# Insert exactly one character into s to get t.
# Delete exactly one character from s to get t.
# Replace exactly one character of s with a different character to get t.

class Solution:
    def isOneEditDistance(self, s: str, t: str) -> bool:
        if s == t:
            return False
        sLen, tLen = len(s), len(t)
        if sLen > tLen: # ensure sLen < tLen
            return Solution.isOneEditDistance(self, t, s)
        if sLen + 1 < tLen:
            return False

        for i in range(tLen):
            if i < sLen and s[i] != t[i] and tLen == sLen:
                return s[:i] + s[i+1:] == t[:i] + t[i+1:]
            elif (i >= sLen or s[i] != t[i]) and s == t[:i] + t[i+1:]:
                # tLen is longer
                return True
        return False